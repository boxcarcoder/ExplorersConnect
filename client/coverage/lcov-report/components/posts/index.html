
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for components/posts</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../index.html">All files</a> components/posts</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">48.39% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>30/62</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">40.74% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>11/27</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">30.43% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>7/23</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">50.85% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>30/59</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
    </div>
    <div class='status-line low'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="Comment.js"><a href="Comment.js.html">Comment.js</a></td>
	<td data-value="96" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 96%"></div><div class="cover-empty" style="width: 4%"></div></div>
	</td>
	<td data-value="96" class="pct high">96%</td>
	<td data-value="25" class="abs high">24/25</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="11" class="abs high">11/11</td>
	<td data-value="87.5" class="pct high">87.5%</td>
	<td data-value="8" class="abs high">7/8</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="24" class="abs high">24/24</td>
	</tr>

<tr>
	<td class="file low" data-value="CommentItem.js"><a href="CommentItem.js.html">CommentItem.js</a></td>
	<td data-value="27.27" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 27%"></div><div class="cover-empty" style="width: 73%"></div></div>
	</td>
	<td data-value="27.27" class="pct low">27.27%</td>
	<td data-value="11" class="abs low">3/11</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="4" class="abs low">0/4</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="5" class="abs low">0/5</td>
	<td data-value="30" class="pct low">30%</td>
	<td data-value="10" class="abs low">3/10</td>
	</tr>

<tr>
	<td class="file low" data-value="PostItem.js"><a href="PostItem.js.html">PostItem.js</a></td>
	<td data-value="11.54" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 11%"></div><div class="cover-empty" style="width: 89%"></div></div>
	</td>
	<td data-value="11.54" class="pct low">11.54%</td>
	<td data-value="26" class="abs low">3/26</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="12" class="abs low">0/12</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="10" class="abs low">0/10</td>
	<td data-value="12" class="pct low">12%</td>
	<td data-value="25" class="abs low">3/25</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank">istanbul</a>
                at Tue Oct 06 2020 19:01:54 GMT-0700 (Pacific Daylight Time)
            </div>
        </div>
        <script src="../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../sorter.js"></script>
        <script src="../../block-navigation.js"></script>
    </body>
</html>
    