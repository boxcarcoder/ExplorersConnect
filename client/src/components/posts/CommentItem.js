import React, { Fragment } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

const CommentItem = ({ comment: { user, avatar, name, text } }) => {
  return (
    <Fragment>
      <div className='post bg-white vert-m-1 p-1'>
        {/*Poster avatar and name */}
        <div>
          <Link to={`/profile/${user}`}>
            <img className='round-img' src={avatar} alt='avatar' />
            <h4>{name}</h4>
          </Link>
        </div>
        {/*Thumbs up, down, and comment */}
        <div>
          <p className='vert-m-1'>{text}</p>
        </div>
      </div>
    </Fragment>
  );
};

CommentItem.propType = {
  postState: PropTypes.object.isRequired
};

const mapStateToProps = state => ({
  postState: state.post
});

export default connect(mapStateToProps)(CommentItem);
