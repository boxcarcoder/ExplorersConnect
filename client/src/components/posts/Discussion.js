import React, { Fragment, useState, useEffect } from 'react';
import { Link } from 'react-router-dom';

import { connect } from 'react-redux';
import { getPost, commentOnPost } from '../../actions/post';

import Spinner from '../layout/Spinner';

import PropTypes from 'prop-types';

const Discussion = ({
  post: { post, loading },
  getPost,
  match,
  commentOnPost
}) => {
  //when discussion component first loads, load redux state with post data
  useEffect(() => {
    getPost(match.params.id);
  }, [getPost, match.params.id]);

  const [formData, setFormData] = useState('');

  const handleSubmit = e => {
    e.preventDefault();
    console.log('commenting on post. ');
    commentOnPost(match.params.id, formData);
  };

  const onChange = e => {
    setFormData({
      [e.target.name]: e.target.value
    });
  };

  if (!post || loading) {
    return <Spinner />;
  } else {
    const { user, avatar, name, text } = post;

    return (
      <Fragment>
        {/*Posts's avatar, name, and text */}
        <div className='post bg-white vert-m-1 p-1'>
          {/*Poster avatar and name */}
          <div>
            <Link to={`/profile/${user}`}>
              <img className='round-img' src={avatar} alt='avatar' />
              <h4>{name}</h4>
            </Link>
          </div>
          {/* Comment */}
          <div>
            <p className='vert-m-1'>{text}</p>
            <Link to='/posts' className='btn btn-primary'>
              Go Back
            </Link>
          </div>
        </div>

        {/*Comment input box */}
        <div className='post-form'  onSubmit={e => handleSubmit(e)}>
          <div className='bg-primary post-form-header '>
            <h4>Leave a Comment</h4>
          </div>
          <form className='form vert-m-1'>
            <textarea
              cols='30'
              rows='5'
              placeholder='Comment on this post'
              name='text'
              onChange={e => onChange(e)}
            ></textarea>
            <input
              type='submit'
              value='Submit'
              className='btn btn-dark vert-m-1'
             
            />
          </form>
        </div>
      </Fragment>
    );
  }
};

Discussion.propTypes = {
  post: PropTypes.object.isRequired,
  getPost: PropTypes.func.isRequired,
  commentOnPost: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  post: state.post
});

export default connect(mapStateToProps, { getPost, commentOnPost })(Discussion);
