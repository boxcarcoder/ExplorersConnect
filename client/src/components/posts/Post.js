import React, { Fragment } from 'react';
import { Link } from 'react-router-dom';

/*need to get profile id for each post. need post component for each post*/
const Post = ({ post }) => {
  const { text, name, avatar, user, likes, comments } = post;

  return (
    <Fragment>
      <div className='post bg-white vert-m-1 p-1'>
        <div>
          <Link to={`/profile/${user}`}>
            <img className='round-img' src={avatar} alt='avatar' />
            <h4>{name}</h4>
          </Link>
        </div>
        <div>
          <p className='vert-m-1'>{text}</p>
          <button className='btn'>
            <i className='fas fa-thumbs-up'></i> <span>4</span>
          </button>
          <button className='btn'>
            <i className='fas fa-thumbs-down'></i>
          </button>
          <Link to='/post' className='btn btn-primary'>
            Discussion
          </Link>
        </div>
      </div>
    </Fragment>
  );
};

export default Post;
