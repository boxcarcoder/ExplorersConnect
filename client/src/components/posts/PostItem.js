import React, { Fragment } from 'react';
import { Link } from 'react-router-dom';

import { likePost, unlikePost, deletePost } from '../../actions/post';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import Spinner from '../layout/Spinner';

import { setAlert } from '../../actions/alert';

const PostItem = ({
  post: { _id, text, name, avatar, user, likes, comments },
  postState: { loading },
  likePost,
  authState: { isAuthenticated, loggedInUser },
  setAlert,
  unlikePost,
  showCommentBtn,
  deletePost
}) => {
  const handleLike = e => {
    if (isAuthenticated) {
      likePost(_id);
    } else {
      setAlert('Please log in to like or dislike post.', 'danger');
    }
  };

  const handleUnlike = e => {
    if (isAuthenticated) {
      unlikePost(_id);
    } else {
      setAlert('Please log in to like or dislike post.', 'danger');
    }
  };

  const displayCommentBtn = () => {
    if (showCommentBtn) {
      return (
        <Link
          to={`/posts/${_id}`}
          className='btn btn-primary btn-small vert-m-1'
        >
          Comment ({comments.length})
        </Link>
      );
    }
  };

  const handleDeletePost = e => {
    deletePost(_id);
  };

  const displayDeleteBtn = () => {
    if (isAuthenticated && loggedInUser._id === user) {
      return (
        <button
          className='btn btn-danger btn-small vert-m-1'
          onClick={e => handleDeletePost(e)}
        >
          x
        </button>
      );
    }
  };

  // //for async operation. Need to wait for post redux state to be populated before proceeding.
  if (loading) {
    return <Spinner />;
  } else {
    return (
      <Fragment>
        <div className='post bg-white vert-m-1 p-1'>
          {/*Poster avatar and name */}
          <div>
            <Link to={`/profile/${user}`}>
              <img className='round-img' src={avatar} alt='avatar' />
              <h4>{name}</h4>
            </Link>
          </div>
          {/*Thumbs up, down, comment, and delete button */}
          <div>
            <p className='vert-m-1'>{text}</p>
            <button className='btn' onClick={e => handleLike(e)}>
              <i className='fas fa-thumbs-up'></i> <span>{likes.length}</span>
            </button>
            <button className='btn' onClick={e => handleUnlike(e)}>
              <i className='fas fa-thumbs-down'></i>
            </button>
            <div>
              {displayCommentBtn()}
              {displayDeleteBtn()}
            </div>
          </div>
        </div>
      </Fragment>
    );
  }
};

PostItem.propTypes = {
  post: PropTypes.object.isRequired,
  likePost: PropTypes.func.isRequired,
  setAlert: PropTypes.func.isRequired,
  unlikePost: PropTypes.func.isRequired,
  postState: PropTypes.object.isRequired,
  authState: PropTypes.object.isRequired,
  deletePost: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  postState: state.post,
  authState: state.auth
});

export default connect(mapStateToProps, {
  likePost,
  setAlert,
  unlikePost,
  deletePost
})(PostItem);
