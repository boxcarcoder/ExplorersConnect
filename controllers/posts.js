const postsRouter = require('express').Router();
const { check, validationResult } = require('express-validator');
const auth = require('../middlewares/auth');

const Post = require('../models/Post');
const User = require('../models/User');

// @route   POST api/posts
// @desc    Create a post
// @access  Private since only a logged in user can make a post
postsRouter.post(
  '/',
  [auth, [check('text', 'Please enter text for the post.').not().isEmpty()]],
  async (req, res) => {
    let errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ errors: errors.array() });
    }

    try {
      // retrieve the logged in user from the db
      let user = await User.findById(req.user.id).select('-password');

      // create new post
      let newPost = new Post({
        text: req.body.text, //get the post text from the req
        name: user.name, //get the name and avatar of the logged in user from the db
        avatar: user.avatar,
        user: req.user.id, // get the user ID from the token of the logged in user
      });

      let post = await newPost.save();

      res.json(post);
    } catch (err) {
      res.status(500).json('Server error.');
    }
  }
);

// @route   GET api/posts
// @desc    Fetch all posts
// @access  Public so site visitors can see posts for more exposure
postsRouter.get('/', async (req, res) => {
  try {
    let posts = await Post.find().sort({ date: -1 }); //sort posts by most recent

    res.json(posts);
  } catch (err) {
    res.status(500).json('Server Error.');
  }
});

// @route   GET api/posts/:postID
// @desc    Fetch post by ID
// @access  Public so site visitors can see posts for more exposure
postsRouter.get('/:postID', async (req, res) => {
  try {
    let post = await Post.findById(req.params.postID);

    if (!post) {
      return res.status(404).json('This post could not be found.');
    } else {
      res.json(post);
    }
  } catch (err) {
    if (err.kind === 'ObjectId') {
      //a properly formatted ID was not found in the request
      return res.status(400).status('This post could not be found.');
    }
    res.status(500).json('Server Error.');
  }
});

// @route   DELETE api/posts/:postID
// @desc    Delete post by ID
// @access  Private since only a logged in user can delete their post
postsRouter.delete('/:postID', auth, async (req, res) => {
  try {
    let post = await Post.findById(req.params.postID);

    if (!post) {
      return res.status(404).json('This post could not be found.');
    } else {
      // check if user is the logged in user
      if (req.user.id !== post.user.toString()) {
        return res.status(401).json('User is not authorized.');
      }

      await post.remove();

      res.json('Post removed.');
    }
  } catch (err) {
    if (err.kind === 'ObjectId') {
      return res.status(400).status('This post could not be found.');
    }
    res.status(500).json('Server Error.');
  }
});

// @route   PUT api/posts/like/:postID
// @desc    Like a post
// @access  Private since only a logged in user can like posts
postsRouter.put('/like/:postID', auth, async (req, res) => {
  try {
    let post = await Post.findById(req.params.postID);

    // check if the post has been liked already by the logged in user
    // prettier-ignore
    if (post.likes.filter(like => like.user.toString() === req.user.id).length > 0) {
      return res.status(405).json('Cannot like more than once');
    }
    else {
    // add the user to the post's likes array
      post.likes.unshift({ user: req.user.id });

      await post.save();

      res.json(post.likes);
    }
  } catch (err) {
    res.status(500).json('Server Error.');
  }
});

// @route   PUT api/posts/unlike/:postID
// @desc    Unlike a post
// @access  Private since only a logged in user can unlike posts
postsRouter.put('/unlike/:postID', auth, async (req, res) => {
  try {
    let post = await Post.findById(req.params.postID);

    // check if the post has been liked already by the logged in user
    // prettier-ignore
    if (post.likes.filter(like => like.user.toString() === req.user.id).length === 0 ) {
      return res.status(405).json('Post has not been liked yet.');
    }
    else {
      //get the remove index
      let removeIdx = post.likes
        .map((like) => like.user.toString())
        .indexOf(req.user.id);

      post.likes.splice(removeIdx, 1);

      await post.save();

      res.json(post.likes);
    }
  } catch (err) {
    res.status(500).json('Server Error.');
  }
});

// @route   POST api/posts/comment/:postID
// @desc    Comment on a post
// @access  Private since only a logged in user can comment on a post
postsRouter.post(
  '/comment/:postID',
  [auth, [check('text', 'Please enter text for the post.').not().isEmpty()]],
  async (req, res) => {
    let errors = validationResult(req);
    if (!errors.isEmpty()) {
      res.status(400).json({ errors: errors.array() });
    } else {
      try {
        // retrieve the logged in user from the db
        let user = await User.findById(req.user.id).select('-password');
        // retrieve the post by the post ID
        let post = await Post.findById(req.params.postID);

        // add new comment to the post
        let newComment = {
          text: req.body.text, //get the post text from the req
          name: user.name, //get the name and avatar of the logged in user from the db
          avatar: user.avatar,
          user: req.user.id, // get the user ID from the token of the logged in user
        };
        post.comments.unshift(newComment);

        await post.save();
        res.json(post.comments);
      } catch (err) {
        res.status(500).json('Server error.');
      }
    }
  }
);

// @route   DELETE api/posts/comment/:postID/:commentID
// @desc    Delete a comment from a post
// @access  Private since only a logged in user can delete a comment
postsRouter.delete('/comment/:postID/:commentID', auth, async (req, res) => {
  try {
    let post = await Post.findById(req.params.postID);

    //retrieve the comment
    let comment = post.comments.find(
      (comment) => comment.id === req.params.commentID
    );

    if (!comment) {
      return res.status(404).json('Comment does not exist.');
    } else {
      // check if user is the logged in user
      if (req.user.id !== comment.user.toString()) {
        return res.status(401).json('User is not authorized.');
      }

      //get the remove index
      let removeIdx = post.comments
        .map((comment) => comment.user.toString())
        .indexOf(req.user.id);

      post.comments.splice(removeIdx, 1);

      await post.save();

      res.json(post.comments);
    }
  } catch (err) {
    res.status(500).json('Server error.');
  }
});

module.exports = postsRouter;
