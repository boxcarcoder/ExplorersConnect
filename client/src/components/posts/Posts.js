import React, { Fragment, useEffect, useState } from 'react';
// import { Link } from 'react-router-dom';

import { connect } from 'react-redux';
import { getAllPosts, addPost } from '../../actions/post';

import PropTypes from 'prop-types';

import Post from './Post';

import Spinner from '../layout/Spinner';

const Posts = ({ getAllPosts, addPost, post: { posts, loading } }) => {
  //on initial load, populate post redux state with all posts
  useEffect(() => {
    getAllPosts();
  }, [getAllPosts]);

  // create state to handle input
  const [formData, setFormData] = useState('');

  const handleSubmit = e => {
    console.log('submitting post: ', formData);
    e.preventDefault();
    addPost(formData);
    setFormData('');
  };

  const onChange = e => {
    // console.log('[e.target.name]', [e.target.name]);
    // console.log('e.target.value', e.target.value);
    setFormData({
      [e.target.name]: e.target.value
    });
  };

  const allPosts = () => {
    console.log('all posts: ', posts);
    return posts.map(post => <Post key={post._id} post={post}/>);
  };

  //need to wait for posts to be loaded into the post redux state before proceeding
  if (!posts.length || loading) {
    return (
      <Fragment>
        <Spinner />
      </Fragment>
    );
  } else {
    return (
      <Fragment>
        <h1 className='large text-primary'>Posts</h1>
        <p className='lead'>
          <i className='fas fa-user'></i> Welcome to the community
        </p>
        {/* Post Box and Submit Button */}
        <div className='post-form'>
          <div className='bg-primary post-form-header '>
            <h4>Start a Discussion</h4>
          </div>

          <form className='form vert-m-1' onSubmit={handleSubmit}>
            <textarea
              cols='30'
              rows='5'
              placeholder='Create a post'
              name='text'
              onChange={e => onChange(e)}
            ></textarea>
            <input
              type='submit'
              className='btn btn-primary vert-m-1'
              value='Submit'
            />
          </form>
        </div>
        {/*List of Posts*/}
        <div className='posts'>{allPosts()}</div>
      </Fragment>
    );
  }
};

Posts.propTypes = {
  post: PropTypes.object.isRequired
};

const mapStateToProps = state => ({
  post: state.post
});

export default connect(mapStateToProps, { getAllPosts, addPost })(Posts);
