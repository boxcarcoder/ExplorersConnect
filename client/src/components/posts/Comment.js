import React, { Fragment, useState, useEffect } from 'react';
import { connect } from 'react-redux';
import { getPost, commentOnPost } from '../../actions/post';
import Spinner from '../layout/Spinner';
import PropTypes from 'prop-types';
import { setAlert } from '../../actions/alert';
import PostItem from './PostItem';
import CommentItem from './CommentItem';

const Comment = ({
  postState: { post, loading },
  getPost,
  match,
  commentOnPost,
  authState: { isAuthenticated },
  setAlert
}) => {
  //when comment component first loads, load redux state with post data
  useEffect(() => {
    getPost(match.params.id);
  }, [getPost, match.params.id]);

  const [formData, setFormData] = useState('');

  const handleSubmit = e => {
    e.preventDefault();
    if (isAuthenticated) {
      commentOnPost(match.params.id, formData);
      setAlert('Commented successfully.', 'success');
    } else {
      setAlert('Please log in to comment on post.', 'danger');
    }
  };

  const onChange = e => {
    setFormData({
      [e.target.name]: e.target.value
    });
  };

  if (!post || loading) {
    return <Spinner />;
  } else {
    return (
      <Fragment>
        {/*Posts's avatar, name, and text */}
        <PostItem key={post._id} post={post} />

        {/*Comment input box */}
        <div className='post-form' onSubmit={e => handleSubmit(e)}>
          <div className='bg-primary post-form-header '>
            <h4>Leave a Comment</h4>
          </div>
          <form className='form vert-m-1'>
            <textarea
              cols='30'
              rows='5'
              placeholder='Comment on this post'
              name='text'
              onChange={e => onChange(e)}
            ></textarea>
            <input
              type='submit'
              value='Submit'
              className='btn btn-dark vert-m-1'
            />
          </form>

          {/* Comments list section */}
          <div className='comments'>
            {post.comments.map(comment => (
              <CommentItem
                key={comment._id}
                comment={comment}
                postId={post._id}
              />
            ))}
          </div>
        </div>
      </Fragment>
    );
  }
};

Comment.propTypes = {
  postState: PropTypes.object.isRequired,
  getPost: PropTypes.func.isRequired,
  commentOnPost: PropTypes.func.isRequired,
  authState: PropTypes.object.isRequired,
  setAlert: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  postState: state.post,
  authState: state.auth
});

export default connect(mapStateToProps, {
  getPost,
  commentOnPost,
  setAlert
})(Comment);
