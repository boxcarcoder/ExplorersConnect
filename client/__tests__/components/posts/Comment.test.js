import React from 'react';
import { shallow } from 'enzyme';

// Import the non-connected component
import { Comment } from '../../../src/components/posts/Comment';

// Globals for testing
let props;
let wrapper;
const mockGetPost = jest.fn();
const mockCommentOnPost = jest.fn();
const mockSetAlert = jest.fn();
const mockRouteParam = 1;

describe('<Comment /> component.', () => {
  describe('Successfully renders', () => {
    test('a spinner while the loading property in the Post state is true.', () => {
      // Set post state to render a spinner.
      props = {
        postState: {
          post: null,
          loading: true,
          deletedPost: false,
        },
        getPost: mockGetPost,
        commentOnPost: mockCommentOnPost,
        setAlert: mockSetAlert,
        authState: {
          isAuthenticated: false,
        },
        match: {
          params: {
            id: mockRouteParam,
          },
        },
      };
      wrapper = shallow(<Comment {...props} />);

      // Execute the test.
      const spinner = wrapper.find('Spinner');
      expect(spinner.exists()).toBe(true);
    });

    test('a redirect to the Posts component if the existing post has been deleted.', () => {
      // Set post state to render a redirect to Posts component.
      props = {
        postState: {
          post: null,
          loading: false,
          deletedPost: true,
        },
        getPost: mockGetPost,
        commentOnPost: mockCommentOnPost,
        setAlert: mockSetAlert,
        authState: {
          isAuthenticated: false,
        },
        match: {
          params: {
            id: mockRouteParam,
          },
        },
      };
      wrapper = shallow(<Comment {...props} />);

      // Execute the test.
      const redirect = wrapper.find('Redirect');
      expect(redirect.props().to).toBe('/posts');
    });

    test('the Comment component for the existing post.', () => {
      // Set post state to render the Comment component.
      props = {
        postState: {
          post: {
            user: 'test user',
            text: 'test post text.',
            comments: [
              {
                user: 'test user2',
                text: 'test comment text.',
              },
              {
                user: 'test user3',
                text: 'test comment text2.',
              },
            ],
          },
          loading: false,
          deletedPost: false,
        },
        getPost: mockGetPost,
        commentOnPost: mockCommentOnPost,
        setAlert: mockSetAlert,
        authState: {
          isAuthenticated: false,
        },
        match: {
          params: {
            id: mockRouteParam,
          },
        },
      };
      wrapper = shallow(<Comment {...props} />);

      // Execute the test.
      expect(wrapper).toMatchSnapshot();
    });
  });

  test('Calls the getPost() action on render.', () => {
    // Spy on React's useEffect method so we can mock it.
    let spyUseEffect = jest.spyOn(React, 'useEffect');

    // Wrap the spied-on function into a mock function to control how many times we can call it.
    const mockUseEffect = () => {
      // Re-define the useEffect method to take the callback function f, and call it synchronously.
      spyUseEffect.mockImplementationOnce((f) => f());
    };

    // Render the component to test if the action is fired.
    props = {
      postState: {
        post: {
          user: 'test user',
          text: 'test post text.',
          comments: [
            {
              user: 'test user2',
              text: 'test comment text.',
            },
            {
              user: 'test user3',
              text: 'test comment text2.',
            },
          ],
        },
        loading: false,
        deletedPost: false,
      },
      getPost: mockGetPost,
      commentOnPost: mockCommentOnPost,
      setAlert: mockSetAlert,
      authState: {
        isAuthenticated: false,
      },
      match: {
        params: {
          id: mockRouteParam,
        },
      },
    };

    // Call the useEffect mock function once.
    mockUseEffect();
    wrapper = shallow(<Comment {...props} />);

    expect(mockGetPost).toHaveBeenCalled();
  });

  describe('If the Comment component renders,', () => {
    beforeEach(() => {
      props = {
        postState: {
          post: {
            user: 'test user',
            text: 'test post text.',
            comments: [
              {
                user: 'test user2',
                text: 'test comment text.',
              },
              {
                user: 'test user3',
                text: 'test comment text2.',
              },
            ],
          },
          loading: false,
          deletedPost: false,
        },
        getPost: mockGetPost,
        commentOnPost: mockCommentOnPost,
        setAlert: mockSetAlert,
        authState: {
          isAuthenticated: true,
        },
        match: {
          params: {
            id: mockRouteParam,
          },
        },
      };
      wrapper = shallow(<Comment {...props} />);
    });

    test('it contains a link to the Posts component.', () => {
      const link = wrapper.find('Link');
      expect(link.props().to).toBe('/posts');
    });

    test('it renders the PostItem component.', () => {
      const postItem = wrapper.find('Connect(PostItem)');
      expect(postItem.exists()).toBe(true);
    });

    test('text updates for the comment input box.', () => {
      // Find the textarea within the component
      wrapper.find('textarea').simulate('change', {
        target: {
          name: 'text',
          value: 'test comment on post.',
        },
      });

      // Execute the test
      expect(wrapper.find('textarea').props().value).toBe(
        'test comment on post.'
      );
    });

    describe('when the form is submitted,', () => {
      test('the default event is cancelled.', () => {
        // The first div contains the onSubmit event handler in the Comment component.
        let eventPrevented = false;
        wrapper
          .find('div')
          .at(0)
          .simulate('submit', {
            preventDefault: () => {
              eventPrevented = true;
            },
          });

        // Execute the test.
        expect(eventPrevented).toBe(true);
      });

      test('fire the commentOnPost() action and setAlert() action if the user is authenticated.', () => {
        // The first div contains the onSubmit event handler in the Comment component.
        wrapper
          .find('div')
          .at(0)
          .simulate('submit', {
            preventDefault: () => {}, //simulate the submit request's need for an event object which contains a preventDefault function.
          });

        // Execute the test.
        expect(mockCommentOnPost).toHaveBeenCalled();
        expect(mockSetAlert).toHaveBeenCalled();
      });

      test('fire the setAlert() action if the user is not authenticated.', () => {
        // Set the props so that isAuthenticated is false.
        props = {
          postState: {
            post: {
              user: 'test user',
              text: 'test post text.',
              comments: [
                {
                  user: 'test user2',
                  text: 'test comment text.',
                },
                {
                  user: 'test user3',
                  text: 'test comment text2.',
                },
              ],
            },
            loading: false,
            deletedPost: false,
          },
          getPost: mockGetPost,
          commentOnPost: mockCommentOnPost,
          setAlert: mockSetAlert,
          authState: {
            isAuthenticated: false,
          },
          match: {
            params: {
              id: mockRouteParam,
            },
          },
        };
        wrapper = shallow(<Comment {...props} />);

        // The first div contains the onSubmit event handler in the Comment component.
        wrapper
          .find('div')
          .at(0)
          .simulate('submit', {
            preventDefault: () => {}, //simulate the submit request's need for an event object which contains a preventDefault function.
          });

        // Execute the test.
        expect(mockSetAlert).toHaveBeenCalled();
      });
    });

    test('it renders all comments for the relevant post.', () => {
      let commentItem = wrapper.find('Connect(CommentItem)').at(0);
      expect(commentItem.exists()).toBe(true);

      commentItem = wrapper.find('Connect(CommentItem)').at(1);
      expect(commentItem.exists()).toBe(true);
    });
  });
});
