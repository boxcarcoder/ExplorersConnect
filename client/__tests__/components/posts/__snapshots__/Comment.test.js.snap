// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Comment /> component. Successfully renders the Comment component for the existing post. 1`] = `
<Fragment>
  <Link
    className="btn"
    to="/posts"
  >
    Go Back
  </Link>
  <Connect(PostItem)
    post={
      Object {
        "comments": Array [
          Object {
            "text": "test comment text.",
            "user": "test user2",
          },
          Object {
            "text": "test comment text2.",
            "user": "test user3",
          },
        ],
        "text": "test post text.",
        "user": "test user",
      }
    }
    showCommentBtn={false}
  />
  <div
    className="post-form"
    onSubmit={[Function]}
  >
    <div
      className="bg-primary post-form-header "
    >
      <h4>
        Leave a Comment
      </h4>
    </div>
    <form
      className="form vert-m-1"
    >
      <textarea
        cols="30"
        name="text"
        onChange={[Function]}
        placeholder="Comment on this post"
        rows="5"
        value=""
      />
      <input
        className="btn btn-dark vert-m-1"
        type="submit"
        value="Submit"
      />
    </form>
    <div
      className="comments"
    >
      <Connect(CommentItem)
        comment={
          Object {
            "text": "test comment text.",
            "user": "test user2",
          }
        }
      />
      <Connect(CommentItem)
        comment={
          Object {
            "text": "test comment text2.",
            "user": "test user3",
          }
        }
      />
    </div>
  </div>
</Fragment>
`;

exports[`<Comment /> component. Successfully renders the Comment component if a post exists. 1`] = `
<Fragment>
  <Link
    className="btn"
    to="/posts"
  >
    Go Back
  </Link>
  <Connect(PostItem)
    post={
      Object {
        "comments": Array [
          Object {
            "text": "test comment text.",
            "user": "test user2",
          },
          Object {
            "text": "test comment text2.",
            "user": "test user3",
          },
        ],
        "text": "test post text.",
        "user": "test user",
      }
    }
    showCommentBtn={false}
  />
  <div
    className="post-form"
    onSubmit={[Function]}
  >
    <div
      className="bg-primary post-form-header "
    >
      <h4>
        Leave a Comment
      </h4>
    </div>
    <form
      className="form vert-m-1"
    >
      <textarea
        cols="30"
        name="text"
        onChange={[Function]}
        placeholder="Comment on this post"
        rows="5"
        value=""
      />
      <input
        className="btn btn-dark vert-m-1"
        type="submit"
        value="Submit"
      />
    </form>
    <div
      className="comments"
    >
      <Connect(CommentItem)
        comment={
          Object {
            "text": "test comment text.",
            "user": "test user2",
          }
        }
      />
      <Connect(CommentItem)
        comment={
          Object {
            "text": "test comment text2.",
            "user": "test user3",
          }
        }
      />
    </div>
  </div>
</Fragment>
`;

exports[`<Comment /> component. Successfully renders the Comment component. 1`] = `
<Fragment>
  <Link
    className="btn"
    to="/posts"
  >
    Go Back
  </Link>
  <Connect(PostItem)
    post={
      Object {
        "comments": Array [
          Object {
            "text": "test comment text.",
            "user": "test user2",
          },
          Object {
            "text": "test comment text2.",
            "user": "test user3",
          },
        ],
        "text": "test post text.",
        "user": "test user",
      }
    }
    showCommentBtn={false}
  />
  <div
    className="post-form"
    onSubmit={[Function]}
  >
    <div
      className="bg-primary post-form-header "
    >
      <h4>
        Leave a Comment
      </h4>
    </div>
    <form
      className="form vert-m-1"
    >
      <textarea
        cols="30"
        name="text"
        onChange={[Function]}
        placeholder="Comment on this post"
        rows="5"
        value=""
      />
      <input
        className="btn btn-dark vert-m-1"
        type="submit"
        value="Submit"
      />
    </form>
    <div
      className="comments"
    >
      <Connect(CommentItem)
        comment={
          Object {
            "text": "test comment text.",
            "user": "test user2",
          }
        }
      />
      <Connect(CommentItem)
        comment={
          Object {
            "text": "test comment text2.",
            "user": "test user3",
          }
        }
      />
    </div>
  </div>
</Fragment>
`;
